client/cli/derive/src/substrate_cli.rs

```diff
@@ -0,0 +1,175 @@
+				},
+				(*x.right).to_token_stream(),
+			)
+		})
+		.collect::<HashMap<_, _>>();
+
+	let pkg_name = std::env::var("CARGO_PKG_NAME").unwrap().to_token_stream();
+	let impl_name = attrs
+		.remove("impl_name")
+		.unwrap_or_else(|| pkg_name.clone());
+	let impl_version = attrs
+		.remove("impl_version")
+		.unwrap_or_else(|| get_version().to_token_stream());
+	let support_url = attrs
+		.remove("support_url")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: support_url"));
+	let executable_name = attrs
+		.remove("executable_name")
+		.unwrap_or_else(|| pkg_name.clone());
+	let author = attrs.remove("author").unwrap_or_else(|| {
+		std::env::var("CARGO_PKG_AUTHORS")
+			.unwrap_or_default()
+			.to_token_stream()
+	});
+	let description = attrs.remove("description").unwrap_or_else(|| {
+		std::env::var("CARGO_PKG_DESCRIPTION")
+			.unwrap_or_default()
+			.to_token_stream()
+	});
+	let copyright_start_year = attrs
+		.remove("copyright_start_year")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: copyright_start_year"));
+
+	if !attrs.is_empty() {
+		abort_call_site!(
+			"unknown macro parameters: {}",
+			attrs
+				.keys()
+				.map(|x| x.as_str())
+				.collect::<Vec<_>>()
+				.join(", ")
+		);
+	}
+
+	let mut i: ItemImpl = match syn::parse(i) {
+		Ok(x) => x,
+		_ => abort_call_site!("this macro only works on an impl"),
+	};
+
+	let existing_methods = i
+		.items
+		.iter()
+		.filter_map(|x| match x {
+			ImplItem::Method(x) => Some(x.sig.ident.to_string()),
+			_ => None,
+		})
+		.collect::<HashSet<_>>();
+
+	define_if_missing!(existing_methods, i, {
+		fn impl_name() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn impl_version() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn support_url() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn executable_name() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn author() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn description() -> &'static str
+	});
+
+	define_if_missing!(existing_methods, i, {
+		fn copyright_start_year() -> i32
+	});
+
+	quote!(#i)
+}
+
+fn get_platform() -> String {
+	use platforms::*;
+
+	let env_dash = if TARGET_ENV.is_some() { "-" } else { "" };
+
+	format!(
+		"{}-{}{}{}",
+		TARGET_ARCH.as_str(),
+		TARGET_OS.as_str(),
+		env_dash,
+		TARGET_ENV.map(|x| x.as_str()).unwrap_or(""),
+	)
+}
+
+fn get_version() -> String {
+	let impl_commit = std::env::var("VERGEN_SHA_SHORT")
+		.unwrap_or_else(|_| {
+			abort_call_site!("missing env variable VERGEN_SHA_SHORT. You might want to add a build.rs script that runs substrate_build_script_utils::generate_cargo_keys() or provide the macro attribute impl_version yourself (impl_version = \"1.0.0\")")
+		});
+	let commit_dash = if impl_commit.is_empty() { "" } else { "-" };
+
+	format!(
+		"{}{}{}-{}",
+		std::env::var("CARGO_PKG_VERSION").unwrap_or_default(),
```

Comment:
Same

Reply:
Tested with the new implementation and it works:

```
[130] [12:05:30] ~/r/s/b/n/cli cecton-the-revenge-of-the-cli > c run -- --version
   Compiling node-cli v2.0.0-alpha.5 (/home/cecile/repos/substrate/bin/node/cli)
    Finished dev [unoptimized + debuginfo] target(s) in 53.38s
     Running `/home/cecile/repos/substrate/target/debug/substrate --version`
substrate 2.0.0-alpha.5-41295e618-x86_64-linux-gnu

[0] [12:09:09] ~/r/s/b/n/cli cecton-the-revenge-of-the-cli > c run -- --version
error: no matching package named `node-cli` found
location searched: /home/cecile/repos/substrate/bin/node/cli
prerelease package needs to be specified explicitly
node-cli = { version = "2.0.1-alpha.5" }
required by package `chain-spec-builder v2.0.0-alpha.5 (/home/cecile/repos/substrate/bin/utils/chain-spec-builder)`
[101] [12:09:14] ~/r/s/b/n/cli cecton-the-revenge-of-the-cli > c run -- --version
   Compiling node-runtime v2.0.0-alpha.5 (/home/cecile/repos/substrate/bin/node/runtime)
   Compiling node-cli v2.0.1-alpha.5 (/home/cecile/repos/substrate/bin/node/cli)
   Compiling node-rpc v2.0.0-alpha.5 (/home/cecile/repos/substrate/bin/node/rpc)
   Compiling node-executor v2.0.0-alpha.5 (/home/cecile/repos/substrate/bin/node/executor)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 38s
     Running `/home/cecile/repos/substrate/target/debug/substrate --version`
substrate 2.0.1-alpha.5-41295e618-x86_64-linux-gnu
```